### PR TITLE
This fixes the reconnect loop that occurs after heartbeat timeout.

### PR DESCRIPTION
**Pull request by bdimcheff**
The bug is something like this:
1. Establish a socket.io connection.
2. Kill your internet connection for long enough for the heartbeat
   timeout to fire.
3. Reconnect your internet. The socket will reconnect, but will soon be
   disconnected by a heartbeat timeout from the PREVIOUS websocket
   connection.

This is what happens internally:
- Heartbeat calls transport.onClose
- transport.onClose calls socket.onDisconnect and thus socket.reconnect
- socket.reconnect swaps out the old transport for a new one
- socket.onDisconnect also calls transport.close and thus
  websocket.close
- websocket.close puts the websocket into the CLOSING state
- new transport/websocket successfully establishes a connection
- old websocket (after some time) goes from CLOSING to CLOSED and fires
  its onclose handler which calls transport.onClose on the OLD transport
- old transport still has a reference to the current socket, so it calls
  socket.onDisconnect, which closes the _new_ socket
- loop back and do it all again

In order for this to happen, the reconnect has to take place _after_ the
socket reconnects, but _before_ the original socket times out and
switches to the CLOSED state.

Fix:

My fix is pretty terrible.  I'm checking to make sure the socket's
transport is the same as the transport handling the onClose event.
Perhaps a better solution would be to remove the onclose handler on the
websocket if the connection is explicitly closed.  There also may be
some better way of cleaning up the old transport during a reconnect.

### CODE REVIEW

File: lib/transport.js [hunk 1]
```diff
@@ -217,9 +217,12 @@
       self.open();
     }, this.socket.options['reopen delay']);*/
 
-    this.isOpen = false;
-    this.socket.onClose();
-    this.onDisconnect();
+    /* only call close if transport hasn't been swapped out already */
```
Comment (jaminben) [inline lib/transport.js:220]: maybe add a comment linking to: https://github.com/olark/socket.io-client/pull/1 so we can read your other ideas.

Comment (bdimcheff) [inline lib/transport.js:220]: wat?  It's in the commit message...

**Comment by jaminben** [comment]
ha, so it is, :shipit:

File: dist/socket.io.js [hunk 2]
```diff
@@ -3865,7 +3868,4 @@ var swfobject=function(){var D="undefined",r="object",S="Shockwave Flash",W="Sho
   , this
 );
 
-if (typeof define === "function" && define.amd) {
```
Comment (mjpizz) [inline dist/socket.io.js:3868]: why kill the amd compat?

**Comment by mjpizz** [comment]
:shipit:

File: dist/socket.io.js (hunk 2 above)
Comment (bdimcheff) [inline dist/socket.io.js:3868]: I killed nothing, the build tool did this on its own.

**Comment by bdimcheff** [comment]
This is what we're running in production... I guess I should merge it to master, but I'm not sure what the amd thing is all about.

**Comment by bdimcheff** [comment]
oh, I would say that it is this: https://github.com/olark/socket.io-client/commit/b5585aa09aea75cf65d4fe84096a45f4055c7695

**Pull request merged**